crates/ruff_linter/src/rules/flake8_bugbear/rules/loop_iterator_mutation.rs

```diff
@@ -0,0 +1,270 @@
+            format!("Mutation to loop iterable `{name}` during iteration")
+        } else {
+            format!("Mutation to loop iterable during iteration")
+        }
+    }
+}
+
+/// B909
+pub(crate) fn loop_iterator_mutation(checker: &mut Checker, stmt_for: &StmtFor) {
+    let StmtFor {
+        target: _,
+        iter,
+        body,
+        orelse: _,
+        is_async: _,
+        range: _,
+    } = stmt_for;
+
+    if !matches!(iter.as_ref(), Expr::Name(_) | Expr::Attribute(_)) {
+        return;
+    }
+
+    // Collect mutations to the iterable.
+    let mutations = {
+        let mut visitor = LoopMutationsVisitor::new(iter);
+        visitor.visit_body(body);
+        visitor.mutations
+    };
+
+    // Create a diagnostic for each mutation.
+    for mutation in mutations.values().flatten() {
+        let name = UnqualifiedName::from_expr(iter)
+            .map(|name| name.to_string())
+            .map(SourceCodeSnippet::new);
+        checker
+            .diagnostics
+            .push(Diagnostic::new(LoopIteratorMutation { name }, *mutation));
+    }
+}
+
+/// Returns `true` if the method mutates when called on an iterator.
+fn is_mutating_function(function_name: &str) -> bool {
+    matches!(
+        function_name,
+        "append"
+            | "sort"
+            | "reverse"
+            | "remove"
+            | "clear"
+            | "extend"
+            | "insert"
+            | "pop"
+            | "popitem"
+            | "setdefault"
+            | "update"
+            | "intersection_update"
+            | "difference_update"
+            | "symmetric_difference_update"
+            | "add"
+            | "discard"
+    )
+}
+
+/// A visitor to collect mutations to a variable in a loop.
+#[derive(Debug, Clone)]
+struct LoopMutationsVisitor<'a> {
+    name: &'a Expr,
+    mutations: HashMap<u8, Vec<TextRange>>,
+    branches: Vec<u8>,
+    branch: u8,
+}
+
+impl<'a> LoopMutationsVisitor<'a> {
+    /// Initialize the visitor.
+    fn new(name: &'a Expr) -> Self {
+        Self {
+            name,
+            mutations: HashMap::new(),
+            branches: vec![0],
+            branch: 0,
+        }
+    }
+
+    /// Register a mutation.
+    fn add_mutation(&mut self, range: TextRange) {
+        self.mutations.entry(self.branch).or_default().push(range);
+    }
+
+    /// Handle, e.g., `del items[0]`.
+    fn handle_delete(&mut self, range: TextRange, targets: &'a [Expr]) {
+        for target in targets {
+            if let Expr::Subscript(ExprSubscript {
+                range: _,
+                value,
+                slice: _,
+                ctx: _,
+            }) = target
+            {
+                if ComparableExpr::from(self.name) == ComparableExpr::from(value) {
+                    self.add_mutation(range);
+                }
+            }
+        }
+    }
+
+    /// Handle, e.g., `items[0] = 1`.
+    fn handle_assign(&mut self, range: TextRange, targets: &[Expr]) {
```

Comment:
All these `&Vec<Expr>` were changed to `&[Expr]`. You almost always want the latter -- it's a lot more flexible, because any reference to a vector can be passed in, but so can other kinds of slices.